src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+// Rust Bitcoin Library
+// Written in 2021 by
+//   The rust-bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! BIP341 Implementation
+//!
+//! Implementation of [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki)
+//! signature message
+//!
+
+use consensus::Encodable;
+use hashes::{sha256, Hash};
+use std::{fmt, io};
+use util::taproot::{TapLeafHash, TapSighashHash};
+use {Script, SigHashType, Transaction, TxOut};
+
+/// Contains outputs of previous transactions to provide to the [SigHashCache::signature_hash]
+/// method, In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::Anyone] may be provided
+pub enum Prevouts {
+    /// When modifier `ANYONECANPAY` is provided, only prevout of the current input is needed
+    Anyone(usize, TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so he can reuse
+    /// the same data structure for multiple inputs
+    All(Vec<TxOut>),
+}
+
+const LEAF_VERSION_TAPSCRIPT: u8 = 0xc0;
+
+/// Information related to the script path spending
+pub struct ScriptPath {
+    script: Script,
+    code_separator: i32,
```

Comment:
nit: slight preference to using `u32` because it is consistent with BIP342 description and c++ codebase.

Reply:
Preference for `code_separator_pos` over `code_separtor`.

Reply:
> nit: slight preference to using `u32` because it is consistent with BIP342 description and c++ codebase.

in the python test code the default value is `-1` that's why I used i32.

are you suggesting to change to `u32` and use a default value of `0xFFFFFFFF` ?

Reply:
Yeah, from BIP342

```
codesep_pos (4): the opcode position of the last executed OP_CODESEPARATOR before the currently 
executed signature opcode, with the value in little-endian (or 0xffffffff if none executed)
```

Reply:
renamed and changed type in e9d3777